File: mars_rover_kotlin/src/main/kotlin/MarsRover.kt
package org.example

class MarsRover {
    fun execution(movement: String = "FFF"): String {
        val moves = movement.split("")
        var gridX = 0
        var gridY = 0
        var compass = "N"
        val compassHistory = mutableListOf<String>()

        for (element in moves) {
            if (compassHistory.size > 0) {

                if ((compassHistory[compassHistory.size - 1] == "W" && element == "R") || (compassHistory[compassHistory.size - 1] == "E" && element == "L")) {
                    compass = "S"
                } else if ((compassHistory[compassHistory.size - 1] == "S" && element == "R") || (compassHistory[compassHistory.size - 1] == "N" && element == "L")) {
                    compass = "E"
                } else if ((compassHistory[compassHistory.size - 1] == "E" && element == "R") || (compassHistory[compassHistory.size - 1] == "W" && element == "L")) {
                    compass = "N"
                } else if ((compassHistory[compassHistory.size - 1] == "N" && element == "R") || (compassHistory[compassHistory.size - 1] == "S" && element == "L")) {
                    compass = "W"
                }
            }

            if (compass == "N" && element == "F") {
                gridY += 1
            } else if (compass == "S" && element == "F") {
                gridY -= 1
            } else if (compass == "W" && element == "F") {
                gridX += 1
            } else if (compass == "E" && element == "F") {
                gridX -= 1
            }

            compassHistory.add(compass)
        }

        println("compassHistory: $compassHistory")

        return "$gridX:$gridY:$compass"
    }
}

File: mars_rover/mars_rover.js
const direction = ["R", "L"];

function marsRover(startDirection, dimenstions, movement) {
  const moves = movement.split("");
  let gridX = 0;
  let gridY = 0;
  let compass = "N";
  const dimX = dimenstions[0];
  const dimY = dimenstions[1];
  let compassHistory = [];

  for (let index = 0; index < moves.length; index++) {
    const element = moves[index];

    if (
      (compassHistory[compassHistory.length - 1] === "W" && element === "R") ||
      (compassHistory[compassHistory.length - 1] === "E" && element === "L")
    ) {
      compass = "S";
    } else if (
      (compassHistory[compassHistory.length - 1] === "S" && element === "R") ||
      (compassHistory[compassHistory.length - 1] === "N" && element === "L")
    ) {
      compass = "E";
    } else if (
      (compassHistory[compassHistory.length - 1] === "E" && element === "R") ||
      (compassHistory[compassHistory.length - 1] === "W" && element === "L")
    ) {
      compass = "N";
    } else if (
      (compassHistory[compassHistory.length - 1] === "N" && element === "R") ||
      (compassHistory[compassHistory.length - 1] === "S" && element === "L")
    ) {
      compass = "W";
    }

    if ((compass === "N" && element === "F") || (compass === "W" && element === "F")) {
      gridY += 1;
    } else if ((compass === "S" && element === "F") || compass === "E" && element === "F") {
      gridY -= 1;
    }

    compassHistory.push(compass);
  }
  console.log("compassHistory", compassHistory);

  if (gridX === dimX || gridX === -1 || gridY === dimY || gridY === -1) {
    return "RIP";
  }

  return `${gridX}:${gridY}:${compass}`;
}

// return marsRover([3, 3], "FFR");

module.exports = marsRover;

// ["F", "F"]
